src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSOutputFormatProvider.java

```diff
@@ -0,0 +1,246 @@
+
+    private OutputFormat getDelegateFormatInstance(Configuration configuration) throws IOException {
+      if (delegateFormat != null) {
+        return delegateFormat;
+      }
+
+      String delegateClassName = configuration.get(DELEGATE_OUTPUTFORMAT_CLASSNAME);
+      try {
+        delegateFormat = (OutputFormat) ReflectionUtils
+          .newInstance(configuration.getClassByName(delegateClassName), configuration);
+        return delegateFormat;
+      } catch (ClassNotFoundException e) {
+        throw new IOException(
+          String.format("Unable to instantiate output format for class %s", delegateClassName),
+          e);
+      }
+    }
+
+    @Override
+    public RecordWriter<NullWritable, StructuredRecord> getRecordWriter(TaskAttemptContext taskAttemptContext) throws
+      IOException, InterruptedException {
+      RecordWriter originalWriter = getDelegateFormatInstance(taskAttemptContext.getConfiguration())
+        .getRecordWriter(taskAttemptContext);
+      return new GCSRecordWriter(originalWriter);
+    }
+
+    @Override
+    public void checkOutputSpecs(JobContext jobContext) throws IOException, InterruptedException {
+      getDelegateFormatInstance(jobContext.getConfiguration()).checkOutputSpecs(jobContext);
+    }
+
+    @Override
+    public OutputCommitter getOutputCommitter(TaskAttemptContext taskAttemptContext) throws IOException,
+      InterruptedException {
+      OutputCommitter delegateCommitter = getDelegateFormatInstance(taskAttemptContext.getConfiguration())
+        .getOutputCommitter(taskAttemptContext);
+      return new GCSOutputCommitter(delegateCommitter);
+    }
+  }
+
+  /**
+   * OutputCommitter for GCS
+   */
+  public static class GCSOutputCommitter extends OutputCommitter {
+
+    private final OutputCommitter delegate;
+
+    public GCSOutputCommitter(OutputCommitter delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public void setupJob(JobContext jobContext) throws IOException {
+      delegate.setupJob(jobContext);
+    }
+
+    @Override
+    public void cleanupJob(JobContext jobContext) throws IOException {
+      delegate.cleanupJob(jobContext);
+    }
+
+    @Override
+    public void commitJob(JobContext jobContext) throws IOException {
+      delegate.commitJob(jobContext);
+    }
+
+    @Override
+    public void abortJob(JobContext jobContext, JobStatus.State state) throws IOException {
+      delegate.abortJob(jobContext, state);
+    }
+
+    @Override
+    public void setupTask(TaskAttemptContext taskAttemptContext) throws IOException {
+      delegate.setupTask(taskAttemptContext);
+    }
+
+    @Override
+    public boolean needsTaskCommit(TaskAttemptContext taskAttemptContext) throws IOException {
+      return delegate.needsTaskCommit(taskAttemptContext);
+    }
+
+    @Override
+    public void commitTask(TaskAttemptContext taskAttemptContext) throws IOException {
+      //Update the metric before committing the file.
```

Comment:
can you add some comments about what you found around why we need to update the metadata before the commit and not after

Reply:
Done.